tests/rptest/scale_tests/many_topics_test.py

```diff
@@ -0,0 +1,1296 @@
+        Simulates a cluster upgrade by doing a rolling restart of all started nodes.
+        Each node is put in maintenance mode and is restarted only after all leadership
+        has been drained from it.
+        """
+        def restart_node(node):
+            if safe:
+                self._enable_maintenance_mode(node)
+            self.redpanda.restart_nodes(node)
+            if safe:
+                self._disable_maintenance_mode(node)
+                self._wait_for_leadership_balanced()
+
+        for node in self.redpanda.started_nodes():
+            self.logger.debug(f"Starting restart for node {node.name}")
+            restart_node(node)
+
+    def _stage_create_topics(self, profile):
+        # Brokers list suitable for script arguments
+        brokers = ",".join(self.redpanda.brokers_list())
+
+        # With current settings, there should be single available node
+        node = self.cluster.alloc(ClusterSpec.simple_linux(1))[0]
+
+        # Call function to create the topics
+        topic_details = self._create_many_topics(
+            brokers,
+            node,
+            profile.topic_name_prefix,
+            profile.topic_count,
+            profile.batch_size,
+            profile.num_partitions,
+            profile.num_replicas,
+            profile.use_kafka_batching,
+            topic_name_length=profile.topic_name_length,
+            skip_name_randomization=False)
+
+        # Free node that used to create topics
+        self.cluster.free_single(node)
+
+        return topic_details
+
+    def _wait_workload_progress(self):
+        def _check():
+            metrics = []
+            for node in self._swarm_producers:
+                metrics.append(node.get_metrics_summary(seconds=20).p50)
+            total_rate = sum(metrics)
+            _m = [str(m) for m in metrics]
+            self.logger.debug(f"...last 20 sec rate is {total_rate} "
+                              f"({', '.join(_m)})")
+            return total_rate >= target_rate
+
+        # Value for progress checks is 20 sec
+        # We'll have exactly one node here, so the rate should be exactly as configured
+        target_rate = 1
+        # Safely try to pull the current profile rate
+        if self._current_profile is not None:
+            target_rate = self._current_profile.messages_per_second_per_producer
+
+        self.redpanda.wait_until(
+            _check,
+            timeout_sec=self.PROGRESS_TIMEOUT,
+            backoff_sec=5,
+            err_msg="Producer Swarm nodes not making progress")
+
+    def _wait_for_leadership_stabilized(self, maintenance_node):
+        def is_stabilized(threshold=0.1):
+            leaders_per_node = [
+                self.redpanda.metric_sum('vectorized_cluster_partition_leader',
+                                         nodes=[n]) for n in nodes
+            ]
+            stddev = numpy.std(leaders_per_node)
+            error = stddev / (self._get_partition_count() / len(nodes))
```

Comment:
similar code exists in the MPT already.

Reply:
That code is operated on a per-topic basis and will run very long for 40k. This one is faster for our cause.